src/lib/operators/table_scan/column_vs_column_table_scan_impl.cpp

```diff
@@ -28,66 +29,99 @@ std::string ColumnVsColumnTableScanImpl::description() const { return "ColumnVsC
-       * each segment iterator type (value, value-non-null, dictionary-simd, ...)!
-       * However, not all combinations are valid or possible.
-       * Only data segments (value, dictionary) or reference segments will be compared, as a table with both data and
-       * reference segments is ruled out. Moreover it is not possible to compare strings to any of the four numerical
-       * data types. Therefore, we need to check for these cases and exclude them via the constexpr-if which
-       * reduces the number of combinations.
-       */
-
-      constexpr auto LEFT_IS_REFERENCE_SEGMENT =
-          std::is_same<LeftSegmentIterableType, ReferenceSegmentIterable<LeftType>>{};
-      constexpr auto RIGHT_IS_REFERENCE_SEGMENT =
-          std::is_same<RightSegmentIterableType, ReferenceSegmentIterable<RightType>>{};
-
-      constexpr auto NEITHER_IS_REFERENCE_SEGMENT = !LEFT_IS_REFERENCE_SEGMENT && !RIGHT_IS_REFERENCE_SEGMENT;
-      constexpr auto BOTH_ARE_REFERENCE_SEGMENTS = LEFT_IS_REFERENCE_SEGMENT && RIGHT_IS_REFERENCE_SEGMENT;
-
-      constexpr auto LEFT_IS_STRING_COLUMN = (std::is_same<LeftType, std::string>{});
-      constexpr auto RIGHT_IS_STRING_COLUMN = (std::is_same<RightType, std::string>{});
-
-      constexpr auto NEITHER_IS_STRING_COLUMN = !LEFT_IS_STRING_COLUMN && !RIGHT_IS_STRING_COLUMN;
-      constexpr auto BOTH_ARE_STRING_COLUMNS = LEFT_IS_STRING_COLUMN && RIGHT_IS_STRING_COLUMN;
-
-      if constexpr ((NEITHER_IS_REFERENCE_SEGMENT || BOTH_ARE_REFERENCE_SEGMENTS) &&
-                    (NEITHER_IS_STRING_COLUMN || BOTH_ARE_STRING_COLUMNS)) {
-        // Dirty hack to avoid https://gcc.gnu.org/bugzilla/show_bug.cgi?id=86740
-        const auto left_it_copy = left_it;
-        const auto left_end_copy = left_end;
-        const auto right_it_copy = right_it;
-        const auto chunk_id_copy = chunk_id;
-        const auto& matches_out_ref = matches_out;
-        const auto condition = _predicate_condition;
-
-        with_comparator(condition, [&](auto predicate_comparator) {
-          auto comparator = [predicate_comparator](const auto& left, const auto& right) {
+  using LeftType = typename LeftIterable::ValueType;
+  using RightType = typename RightIterable::ValueType;
+
+  // C++ cannot compare strings and non-strings out of the box:
+  if constexpr (std::is_same_v<LeftType, std::string> == std::is_same_v<RightType, std::string>) {
+    bool flipped = false;
+    auto condition = _predicate_condition;
+    if (condition == PredicateCondition::GreaterThan || condition == PredicateCondition::GreaterThanEquals) {
+      condition = flip_predicate_condition(condition);
+      flipped = true;
+    }
+
+    auto erase_comparator_type = [](auto comparator, const auto& it1, const auto& it2) {
+      if constexpr (type_erasure == SegmentIterationTypeErasure::Always) {
+        return comparator;
+      } else {
+        return std::function<bool(const AbstractSegmentPosition<std::decay_t<decltype(it1->value())>>&,
```

Comment:
Does wrapping it into std::function have any performance impact?

Reply:
Yes, because it type-erases the lambda implementation. This is one level of indirection, but more importantly, breaks the basic block and hinders inlining.